### PR TITLE
feat: module `avm/res/network/virtual-network-gateway`

### LINKED ISSUE #3061: [AVM Module Issue]: Public IP in `virtual-network-gateway` module missing properties
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

I'm not sure

### Module Name

avm/res/network/virtual-network-gateway

### (Optional) Module Version

latest

### Description

Currently, I am busy building a platform landing zone using AVM modules. This landing zone requires a virtual network gateway, so I decided to use this AVM module. The module also deploys the necessary Public IP resource, but it is missing properties such as the DNS domain label, which I need.

Although it’s easy to add these properties to the template, we might need to consider a different approach. According to the documentation, a resource module should not deploy external dependencies. It depends on how we view the module; the Public IP resource is required for the virtual network gateway but is also an external dependency.

As a suggestion, I would recommend using the required Public IP resourceId parameter route. This approach enforces that the Public IP resource must be provided through a parameter. If we want to keep the Public IP and gateway together, we should expose every Public IP property and offer the user the option to provide a resourceId as an alternative to "bring your own IP".

Tagging @AlexanderSehr @fabmas 

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by fabmas**
## Description
Applied the same approach we did for several other modules that may or must use a public IP, allowing the use of an existing Public IP (resourceID)

Closes #3061

## Pipeline Reference

[avm.res.network.virtual-network-gateway](https://github.com/fabmas/bicep-registry-modules/actions/workflows/avm.res.network.virtual-network-gateway.yml)

| Pipeline |
| -------- |
|          |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Review by eriqua — COMMENTED** [review]
Hey @fabmas please merge latest back to this branch as many conflicts popped up

**Comment by ReneHezser** [comment]
@fabmas 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

**Comment by fabmas** [comment]
@ReneHezser I ran the script but files seems already updated because no changes were detected.

**Comment by ReneHezser** [comment]
@fabmas you need to update your branch with the changes that have been applied to the main branch of the brm repository to have the updated version of the script.

**Comment by fabmas** [comment]
closed by mistake

File: avm/res/network/virtual-network-gateway/tests/e2e/vpn-no-az/main.test.bicep [hunk 1]
```diff
@@ -56,9 +56,8 @@ module testDeployment '../../../main.bicep' = [
       name: '${namePrefix}${serviceShort}001'
       skuName: 'VpnGw1'
       gatewayType: 'Vpn'
-      publicIpZones: []
       vNetResourceId: nestedDependencies.outputs.vnetResourceId
-      clusterSettings: {
+      clusterSettings:{
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/tests/e2e/vpn-no-az/main.test.bicep:60]:
```suggestion
      clusterSettings: {
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 2]
```diff
@@ -8,20 +8,19 @@ param name string
 @description('Optional. Location for all resources.')
 param location string = resourceGroup().location
 
-@description('Optional. Specifies the name of the Public IP used by the Virtual Network Gateway. If it\'s not provided, a \'-pip\' suffix will be appended to the gateway\'s name.')
-param gatewayPipName string = '${name}-pip1'
+@description('Optional. The Public IP resource ID to associate to the Virtual Network Gateway. If empty, then the Public IP that is created as part of this module will be applied to the Virtual Network Gateway.')
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:11]:
```suggestion
@description('Optional. The Public IP resource ID to associate to the Virtual Network Gateway. If empty, then a new Public IP will be created and applied to the Virtual Network Gateway.')
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 3]
```diff
@@ -8,20 +8,19 @@ param name string
...
 @description('Optional. Resource ID of the Public IP Prefix object. This is only needed if you want your Public IPs created in a PIP Prefix.')
 param publicIPPrefixResourceId string = ''
 
 @description('Optional. Specifies the zones of the Public IP address. Basic IP SKU does not support Availability Zones.')
-param publicIpZones array = [
-  1
-  2
-  3
-]
+param publicIpZones array = []
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:21]: Why this change?

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 4]
```diff
@@ -179,7 +186,8 @@ var bgpSettingsVar = isActiveActive
       ]
     }
 
-// Potential IP configurations (active-active vs active-passive)
+
+// Potential IP configurations (Active-Active vs active-passive)
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:190]: If change required I guess it should apply also to Active-Passive
```suggestion
// Potential IP configurations (Active-Active vs Active-Passive)
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 5]
```diff
@@ -188,9 +196,15 @@ var ipConfiguration = isActiveActive
           subnet: {
             id: '${vNetResourceId}/subnets/GatewaySubnet'
           }
+          //Use existing Public IP, new Public IP created in this module
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:199]:
```suggestion
          // Use existing Public IP, new Public IP created in this module
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 6]
```diff
@@ -188,9 +196,15 @@ var ipConfiguration = isActiveActive
...
           publicIPAddress: {
-            id: az.resourceId('Microsoft.Network/publicIPAddresses', gatewayPipName)
+            id: !empty(existingFirstPipResourceId)
+              ? existingFirstPipResourceId
+              : az.resourceId('Microsoft.Network/publicIPAddresses', firstPipName)
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:203]: Would this work?
```suggestion
            id: existingFirstPipResourceId ?? az.resourceId('Microsoft.Network/publicIPAddresses', firstPipName)
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 7]
```diff
@@ -188,9 +196,15 @@ var ipConfiguration = isActiveActive
...
           }
+
+
+
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:207]:
```suggestion
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 8]
```diff
@@ -202,8 +216,12 @@ var ipConfiguration = isActiveActive
           }
           publicIPAddress: {
             id: isActiveActive
-              ? az.resourceId('Microsoft.Network/publicIPAddresses', activeGatewayPipNameVar)
-              : az.resourceId('Microsoft.Network/publicIPAddresses', gatewayPipName)
+              ? !empty(existingSecondPipResourceIdVar)
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:219]: same here if `??` works

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 9]
```diff
@@ -217,7 +235,9 @@ var ipConfiguration = isActiveActive
             id: '${vNetResourceId}/subnets/GatewaySubnet'
           }
           publicIPAddress: {
-            id: az.resourceId('Microsoft.Network/publicIPAddresses', gatewayPipName)
+            id: !empty(existingFirstPipResourceId)
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:238]: same here if `??` works

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 10]
```diff
@@ -557,45 +577,59 @@ type diagnosticSettingType = {
...
-
+  
   @description('Optional. The list of the second custom BGP IP Address (APIPA) peering addresses which belong to IP configuration.')
   secondCustomBgpIpAddresses: string[]?
 }
 
 @discriminator('clusterMode')
 type clusterSettingType = activeActiveNoBgpType | activeActiveBgpType | activePassiveBgpType | activePassiveNoBgpType
+
+
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:635]:
```suggestion
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 11]
```diff
@@ -557,45 +577,59 @@ type diagnosticSettingType = {
...
   @description('Optional. The Autonomous System Number value. If it\'s not provided, a default \'65515\' value will be assigned to the ASN.')
   @minValue(0)
   @maxValue(4294967295)
-  asn: int?
+  asn: int? 
 
   @description('Optional. The list of custom BGP IP Address (APIPA) peering addresses which belong to IP configuration.')
   customBgpIpAddresses: string[]?
-
+
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:627]:
```suggestion

```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 12]
```diff
@@ -557,45 +577,59 @@ type diagnosticSettingType = {
...
+  existingSecondPipResourceId: string?
+  
+  @description('Optional. Specifies the name of the secondary Public IP to be created for the Virtual Network Gateway in the Active-Active mode. This will only take effect if no existing secondary Public IP is provided. If neither an existing secondary Public IP nor this parameter is specified, a new secondary Public IP will be created with a default name, using the gateway\'s name with the \'-pip2\' suffix.')
+  secondPipName: string?
+  
   @description('Optional. The Autonomous System Number value. If it\'s not provided, a default \'65515\' value will be assigned to the ASN.')
   @minValue(0)
   @maxValue(4294967295)
-  asn: int?
+  asn: int?
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:623]:
```suggestion
  asn: int?
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 13]
```diff
@@ -557,45 +577,59 @@ type diagnosticSettingType = {
...
+
 }
 
 type activeActiveNoBgpType = {
+  
   clusterMode: 'activeActiveNoBgp'
 
-  @description('Optional. Specifies the name of the Public IP used by the Virtual Network Gateway when active-active configuration is required. If it\'s not provided, a \'-pip2\' suffix will be appended to the gateway\'s name.')
-  activeGatewayPipName: string?
+  @description('Optional. The secondary Public IP resource ID to associate to the Virtual Network Gateway in the Active-Active mode. If empty, then the secondary Public IP that is created as part of this module will be applied to the Virtual Network Gateway.')
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:589]:
```suggestion
  @description('Optional. The secondary Public IP resource ID to associate to the Virtual Network Gateway in the Active-Active mode. If empty, then a new secondary Public IP will be created as part of this module and applied to the Virtual Network Gateway.')
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 14]
```diff
@@ -557,45 +577,59 @@ type diagnosticSettingType = {
...
 
 type activePassiveBgpType = {
+  
   clusterMode: 'activePassiveBgp'
 
   @description('Optional. The Autonomous System Number value. If it\'s not provided, a default \'65515\' value will be assigned to the ASN.')
   @minValue(0)
   @maxValue(4294967295)
-  asn: int?
+  asn: int?
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:604]:
```suggestion
  asn: int?
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 15]
```diff
@@ -557,45 +577,59 @@ type diagnosticSettingType = {
...
 }
 
 type activeActiveBgpType = {
+  
   clusterMode: 'activeActiveBgp'
 
-  @description('Optional. Specifies the name of the Public IP used by the Virtual Network Gateway when active-active configuration is required. If it\'s not provided, a \'-pip2\' suffix will be appended to the gateway\'s name.')
-  activeGatewayPipName: string?
-
+  @description('Optional. The secondary Public IP resource ID to associate to the Virtual Network Gateway in the Active-Active mode. If empty, then the secondary Public IP that is created as part of this module will be applied to the Virtual Network Gateway.')
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:614]:
```suggestion
  @description('Optional. The secondary Public IP resource ID to associate to the Virtual Network Gateway in the Active-Active mode. If empty, then a new secondary Public IP will be created as part of this module and applied to the Virtual Network Gateway.')
```

File: avm/res/network/virtual-network-gateway/main.bicep [hunk 16]
```diff
@@ -143,18 +142,26 @@ var isBgp = (clusterSettings.clusterMode == 'activeActiveBgp' || clusterSettings
...
+        ? []
+        : [secondPipNameVar]
+    )
+  : concat(
+      !empty(existingFirstPipResourceId)
+        ? []
+        : [firstPipName]
+    )
+
+// Potential BGP configurations (Active-Active vs active-passive)
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/main.bicep:164]:
```suggestion
// Potential BGP configurations (Active-Active vs Active-Passive)
```

File: avm/res/network/virtual-network-gateway/tests/e2e/activePassiveExistingPip/main.test.bicep [hunk 17]
```diff
@@ -1,7 +1,7 @@
 targetScope = 'subscription'
 
-metadata name = 'AAD-VPN'
-metadata description = 'This instance deploys the module with the AAD set of required parameters.'
+metadata name = 'VPN Active Passive with BGP settings using existing Public IP'
+metadata description = 'This instance deploys the module with the VPN Active Passive with APIPA BGP settings.'
```
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/tests/e2e/activePassiveExistingPip/main.test.bicep:4]:
```suggestion
metadata description = 'This instance deploys the module with the VPN Active Passive with APIPA BGP settings and existing primary public IP.'
```

File: avm/res/network/virtual-network-gateway/README.md
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/README.md:1]: Please regenerate after addressing other comments

**Review by eriqua — COMMENTED** [review]
Great job @fabmas 
Just a few suggestions. Main point related to the `publicIpZones` default value change

File: avm/res/network/virtual-network-gateway/main.bicep (hunk 3 above)
Comment (fabmas) [reply avm/res/network/virtual-network-gateway/main.bicep:21]: probably an uncontrolled raptus of copy/paste :)

File: avm/res/network/virtual-network-gateway/README.md
Comment (eriqua) [inline avm/res/network/virtual-network-gateway/README.md:1]: Comments seems all addressed 💪 . 
However the pipeline fails on many deployment validation tests. 
Please check and just let me know once ready for rereview 🙂

Comment (fabmas) [reply avm/res/network/virtual-network-gateway/README.md:1]: Some changes had to be rolled back.
[avm.res.network.virtual-network-gateway](https://github.com/fabmas/bicep-registry-modules/actions/workflows/avm.res.network.virtual-network-gateway.yml)

Comment (eriqua) [reply avm/res/network/virtual-network-gateway/README.md:1]: As mentioned offline, that's not a problem, we can investigate on that next.
Please note I still see a red pipeline badge with multiple tests failing. Could you please check?

**Review by eriqua — APPROVED** [review]
🆗
